CBinding/Compiler/ICompiler.cs

```diff
@@ -30,47 +30,431 @@
+					}      
+
+					sb.Clear ();
+					fileName = "";
+					lineNumber = 0;
+					isWarning = false;
+
+					// in/at CMakeLists.txt:10 (COMMAND):
+					if (line.Contains (" in ")) {
+						Tuple<int, string> t = GetFileAndLine (line, " in ");
+						lineNumber = t.Item1;
+						fileName = t.Item2;
+					} else if (line.Contains (" at ")) {
+						Tuple<int, string> t = GetFileAndLine (line, " at ");
+						lineNumber = t.Item1;
+						fileName = t.Item2;
+					} else {
+						string [] error = line.Split (':');
+						if (!string.IsNullOrEmpty (error.ElementAtOrDefault (1))) {
+							sb.Append (error [1]);
+						}
+					}
+				} else {
+					sb.Append (line);
+				}
+			}
+
+			return results;
+		}
+
+		public static CMakeToolchain GetToolchain (string toolchainName )
+		{
+			OperatingSystem os = Environment.OSVersion;
+			if (os.Platform == PlatformID.Win32Windows || os.Platform == PlatformID.Win32S || os.Platform == PlatformID.WinCE || os.Platform == PlatformID.Win32NT) {
+				if (toolchainName == "VS15 MSBuild Toolchain")
+					return new VS15MSBuildToolchain ();
+				else if (toolchainName == "VS14 MSBuild Toolchain")
+					return new VS14MSBuildToolchain ();
+				else if (toolchainName == "VS12 MSBuild Toolchain")
+					return new VS12MSBuildToolchain ();
+				else if (toolchainName == "Clang Toolchain")
+					return new ClangToolchain ();
+				else return new MinGW32Toolchain ();
+			} else if (os.Platform == PlatformID.Unix)
+					return new UnixMakeToolchain ();
+				else if (os.Platform == PlatformID.MacOSX)
+					return new MacMakeToolchain ();
+			else return new MinGW32Toolchain ();
+
+		}
+
+		/// Use cmake to generate makefiles for this toolchain.
+		public abstract Task<BuildResult> GenerateMakefiles (string projectName, FilePath outputDirectory, ProgressMonitor monitor);
+
+	}	
+
+	[Extension("/CBinding/Toolchains")]
+	public class VS15MSBuildToolchain : CMakeToolchain
+	{
+
+	/// <summary>
+	/// The name of this Toolchain.
+	/// </summary>
+	/// <value>The name.</value>
+		public override string ToolchainName {
+			get {
+				return "VS15 MSBuild Toolchain"; }
+		}
+
+		/// Use cmake to generate makefiles for this toolchain.
+		public override Task<BuildResult> GenerateMakefiles (string projectName, FilePath outputDirectory, ProgressMonitor monitor)
+		{
+			monitor.BeginStep ("Generating build files...");
+			Stream generationResult = ExecuteCommand ("cmake", "../ -G \"Visual Studio 15 2017\"", outputDirectory, monitor);
+			BuildResult results = ParseGenerationResult (generationResult, monitor);
+			monitor.EndStep ();
+
+			string projectToBuild = $"{projectName}.\"sln\""; 
+			monitor.BeginStep ("Building...");
+			Stream buildResult = ExecuteCommand ("msbuild", projectToBuild, outputDirectory, monitor);
+			monitor.EndStep ();
+
+			return Task.FromResult (results);
+		}
+
+	}
+
+	[Extension ("/CBinding/Toolchains")]
+	public class VS14MSBuildToolchain : CMakeToolchain
```

Comment:
You could easily share code between VS15MSBuildToolchain/VS14MSBuildToolchain/VS12MSBuildToolchain using an abstract base class, MSBuildToolchain. The subclasses would just override the ID and name.